TUTORIAL.md

```diff
@@ -0,0 +1,245 @@
+
+###Step 1: Set up your app
+- `bin/rake db:migrate RAILS_ENV=test`
+- `bundle install`
+- `rails s` then visit `localhost:3000`
+- `rake db:seed`
+
+At this point we should have a form that will let us create a post, along with tags that we can select. Now we need a way to add a new tag, so let's add that to our form.
+
+###Step 2: Use fields_for to accept tags
+
+```ruby
+<div class="field">
+  <%= f.fields_for(:tags) do |tag_form| %>
+    <%= tag_form.label :name %>
+    <%= tag_form.text_field :name %>
+  <% end %>
+</div>
+
+```
+####Completed Form
+
+```ruby
+<%= form_for(@post) do |f| %>
+  <% if @post.errors.any? %>
+    <div id="error_explanation">
+      <h2><%= pluralize(@post.errors.count, "error") %> prohibited this post from being saved:</h2>
+      <ul>
+      <% @post.errors.full_messages.each do |msg| %>
+        <li><%= msg %></li>
+      <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+  <div class="field">
+    <%= f.label :name %><br>
+    <%= f.text_field :name %><br>
+    <%= f.label :content %><br>
+    <%= f.text_area :content %><br>
+    Tags:
+  </div>
+
+  <div class="field">
+    <%= f.fields_for(:tags) do |tag_form| %>
+      <%= tag_form.label :name %>
+      <%= tag_form.text_field :name %>
+    <% end %>
+  </div>
+
+  <div class="field">
+    <%= f.collection_check_boxes :tag_ids, Tag.all, :id, :name %><br>
+  </div>
+
+  <div class="actions">
+    <%= f.submit %>
+  </div>
+
+<% end %>
+```
+
+####Before we go any further, let's break down what is happening above. 
+
+Our `form_for` accepts `@post` as an argument, which is an instance of our `Post` class. We then pass it to the variable `f` to build our form.
+
+Let's take a look at `@post` params in our console and see what is being returned. 
+
+
+```ruby
+#<Post:0x007fb3ad9e4e50> {
+            :id => 10,
+          :name => "Hello World!",
+       :content => "We are learning about fields_for!",
+       :user_id => nil,
+    :created_at => Thu, 09 Jul 2015 16:59:41 UTC +00:00,
+    :updated_at => Thu, 09 Jul 2015 16:59:41 UTC +00:00
+}
+```
+When we call `<%= f.label :name %>` in our form above, we are accessing the `:name` key in our hash, which is going to return the name of that post, same thing for `:content`, or any other key that is available in our params.
+
+###Step 3: Set up your PostsContrller
+
+### PostsController
+
+```ruby
+def new
+  @post = Post.new
+  @tag = @post.tags.build
+end
+```
+
+Our controller is going to render our form, but what is this `build` method? Our `build` method was given to use via our model associations, which we will take a look at later. It creates a new object in memory, so that our view can use it to display something. In our case, since we are using `fields_for`, it will display fields for our tags! If you delete `@tag = @post.tags.build` from your controller, your fields would disappear.
```

Comment:
and your fields disappear because build creates *one* object. It's not the calling of build that makes the form helper run. It's the fact that build creates a single instance, which the fields for iterates over. IF you called build, three times, it creates three different objects and then it would happen three times. Showing this causality kinda flow chart is pretty important.

Last, if you take a look, in your `fields_for` are you actually calling `@tags`? So how dos the fields_for know about the new tags?!?! Remember! build creates the associations. What happens is fields_for calls the tags method on the post object. Because you ahve created those thigns with build! It makes the associations in both directions. You can prove this by not calling build but by just creating a blank tag object and then setting up the association manually.